socs/Lakeshore/Lakeshore336.py

```diff
@@ -0,0 +1,1195 @@
+# proxy for agent
+sys.path.append('/home/zatkins/repos/stslab/testbed')
+
+# helper dicts
+sensor_key = {
+    '0': 'disabled',
+    '1': 'diode',
+    '2': 'platinum rtd',
+    '3': 'ntc rtd',
+    '4': 'thermocouple',
+    '5': 'capacitance'
+}
+sensor_lock = {v:k for k, v in sensor_key.items()}
+
+units_key = {
+    '1': 'kelvin',
+    '2': 'celsius',
+    '3': 'sensor'
+}
+units_lock = {v:k for k, v in units_key.items()}
+
+tempco_key = {
+    '1': 'negative',
+    '2': 'positive'
+}
+tempco_lock = {v:k for k, v in tempco_key.items()}
+
+format_key = {
+    '1': "mV/K (linear)",
+    '2': "V/K (linear)",
+    '3': "Ohm/K (linear)",
+    '4': "log Ohm/K (linear)"
+}
+format_lock = {v:k for k, v in format_key.items()}
+
+output_modes_key = {
+    '0': 'off', 
+    '1': 'closed loop', 
+    '2': 'zone', 
+    '3': 'open loop', 
+    '4': 'monitor out', 
+    '5': 'Warm up'
+}
+output_modes_lock = {v.lower():k for k, v in output_modes_key.items()}
+
+channel_key = {
+    '0': 'none',
+    '1': 'A',
+    '2': 'B',
+    '3': 'C',
+    '4': 'D'
+}
+channel_lock = {v:k for k,v in channel_key.items()}
+
+heater_display_key = {
+    '1': 'current',
+    '2': 'power'
+}
+heater_display_lock = {v:k for k,v in heater_display_key.items()}
+
+max_current_key = {
+    '0': 'User',
+    '1': .707,
+    '2': 1.,
+    '3': 1.141,
+    '4': 2
+}
+max_current_lock = {v:k for k, v in max_current_key.items()}
+
+heater_range_key = {
+    "0": "off",
+    "1": "low",
+    "2": "medium",
+    "3": "high"
+}
+heater_range_lock = {v:k for k, v in heater_range_key.items()}
+
+ramp_key = {
+    '0': 'off',
+    '1': 'on'
+}
+ramp_lock = {v:k for k,v in ramp_key.items()}
+
+# main class - Lakeshore 336 driver
+class LS336:
+    '''
+    Implements a lakeshore 336 box to interface with client scripts. Only contains locally relevant information; namely, port parameters. The state of the device is not stored locally to avoid the potential for inconsistent information. Device status can always be accessed (accurately) through a msg.    
+    '''
+    #Constructor and instance variables
+
+    def __init__(self, port, timeout = 1.0):
+
+        # LS336 defaults
+        self.baud = 57600
+        self.bytesize = serial.SEVENBITS
+        self.parity = serial.PARITY_ODD
+        self.stopbits = serial.STOPBITS_ONE
+        self.term = '\r\n'
+        self.std_delay = 0.06
+
+        self.port = port
+        self.timeout = timeout
+        self.comm = serial.Serial(self.port, self.baud, self.bytesize, self.parity, self.stopbits, timeout)
+        self.comm.close()
+
+        self.id = self.get_id()
+        print(self.id) #print idenfitication information to see if working
+
+        # Get Channels
+        inps = ['A','B','C','D']
+        self.channels = {inp:Channel(self, inp) for inp in inps}
+
+        # Get Heaters
+        htrs = ['1', '2']
+        self.heaters = {out:Heater(self, out) for out in htrs}
+
+    #Instance methods
+
+    def msg(self, cmd, delay = None):
```

Comment:
1. I have added an assert statement to the msg method to ensure the delay is greater than 50ms for all messages, including queries
2. This will be true by the above (since 1/0.05 = 20)

1. This is fixed, see previous conversation
2. This is true by default in the pyserial read_until method (it looks for '\n', and LS336 terminates messages with '\r\n')
3. See above
4. See above

Reply:
re: 4. -- maybe `read_until` used to be used, but the call to `self.com.recv()` doesn't ensure the entire response is read, so we should make sure it fits in 4096 bytes. Everything else looks good.